## Applied Data Science Capstone Project
### Week 3 - Segmenting and Clustering Neighborhoods in Toronto
### Kevin Spradlin
### July 12, 2021

## Step 1 - Webscrape Postal Codes, Boroughs, and Neighborhoods

In [1]:
!pip install bs4
!pip install html5lib
#!pip install requests

In [2]:
from bs4 import BeautifulSoup # this module helps in web scrapping.
import requests  # this module helps us to download a web page
import pandas as pd

#### Retrieve the page with the Toronto postal code, borough, and neighborhood table

In [3]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

data  = requests.get(url).text

#soup = BeautifulSoup(data,"html5lib")
soup = BeautifulSoup(data,"html.parser")


In [4]:
#print(soup.prettify())

#find all html tables in the web page
tables = soup.find_all('table') # in html table is represented by the tag <table>

len(tables)


3

#### Find the table with the postal codes.  It's the one with 'Not assigned' in some of its cells.

In [5]:
# find the table with the postal codes
for index,table in enumerate(tables):
    if ("Not assigned" in str(table)):
        table_index = index
print(table_index)

#print(tables[table_index].prettify())


0


#### Extract the table's data into a list, then copy it into a _pandas_ dataframe.  You need to simplify some boroughs' names.

In [6]:
# extract the postal codes, boroughs, and neighborhoods into a list
table_contents = []

for row in tables[table_index].findAll("td"):
  cell = {}

  if row.span.text == "Not assigned":
    pass
  else:
    cell["PostalCode"] = row.p.text[:3]
    cell["Borough"] = (row.span.text).split('(')[0]
    cell["Neighborhood"] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
    table_contents.append(cell)

#print(table_contents)


In [7]:
# convert the list with postal codes, boroughs, and neighborhoods into a pandas dataframe
df = pd.DataFrame(table_contents)

#df.Borough.unique()

df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                     'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                     'EtobicokeNorthwest':'Etobicoke Northwest',
                                     'East YorkEast Toronto':'East York/East Toronto',
                                     'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

#df.Borough.unique()


In [8]:
# print the dimensions of the dataframe
df.shape

(103, 3)

### The dataframe has 103 rows (postal codes for boroughs and neighborhoods).

## Step 2 - Get the latitude and longitude of each postal code

In [9]:
!pip install geocoder
import geocoder


#### I tried to use the geocoder module, but after 2 minutes it was still running.  Next, I tried using geocoder to get the coordinates for one postal code.  It ran for 2 minutes without getting a result.  So, I decided to just use the csv file.

In [11]:
# loop through the rows in the dataframe, putting the combined results into a new dataframe
full_table_contents = []


for index, row in df.iterrows():
  #print(row['PostalCode'])

  cell = {}
  cell["PostalCode"] = row["PostalCode"]  
  cell["Borough"] = row["Borough"]  
  cell["Neighborhood"] = row["Neighborhood"]  


  lat_long_coords = None

  # keep querying until you get coordinates for the postal code
  while lat_long_coords is None:
    geo_query = geocoder.google(f"{row['PostalCode']:s}, Toronto, Ontario")
    lat_long_coords = geo_query.latlng


  cell["Latitude"] = lat_long_coords[0]
  cell["Longitude"] =lat_long_coords[1]

  full_table_contents.append(cell)


full_df = pd.DataFrame(full_table_contents)

full_df.head()


In [11]:
# note - tested geocoder with one postal code.  I didn't get a response after 2 minutes.
lat_long_coords = None

postal_code = "M5A"

while lat_long_coords is None:
  geo_query = geocoder.google("{postal_code:s}, Toronto, Ontario")
  lat_long_coords = geo_query.latlng


print(lat_long_coords)
  

#### Read the data in the csv file into a dictionary, then combine it and the postal code/borough/neighborhood data into a new _toronto_data_ dataframe.

In [12]:
# read the coordinates from the csv file into a dictionary.
lat_long_coords = {}

geodata = open('Geospatial_Coordinates.csv', 'r')

for curr_line in geodata:
  postal_code, latitude, longitude = curr_line.split(',')

  lat_long_coords[postal_code] = (latitude, longitude[:-1], )


geodata.close()


In [13]:
# loop through the rows in the dataframe, putting the combined postal code/borough/neighborhood/latitude and longitude information into a new dataframe
full_table_contents = []


for index, row in df.iterrows():
  cell = {}
  cell["PostalCode"] = row["PostalCode"]  
  cell["Borough"] = row["Borough"]  
  cell["Neighborhood"] = row["Neighborhood"]  

  if row["PostalCode"] in lat_long_coords:
    cell["Latitude"] = lat_long_coords[row["PostalCode"]][0]
    cell["Longitude"] = lat_long_coords[row["PostalCode"]][1]

  full_table_contents.append(cell)


toronto_data = pd.DataFrame(full_table_contents)

toronto_data.head()


PostalCode           Borough                      Neighborhood    Latitude  \
0        M3A        North York                         Parkwoods  43.7532586   
1        M4A        North York                  Victoria Village  43.7258823   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.6542599   
3        M6A        North York  Lawrence Manor, Lawrence Heights   43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.6623015   

     Longitude  
0  -79.3296565  
1  -79.3155716  
2  -79.3606359  
3  -79.4647633  
4  -79.3894938

In [14]:
toronto_data.shape

(103, 5)

### The new dataframe has 103 rows, or the same number as the original dataframe.  So no rows were left out.

In [15]:
toronto_data.groupby('Borough').count()

PostalCode  Neighborhood  Latitude  Longitude
Borough                                                              
Central Toronto                  9             9         9          9
Downtown Toronto                17            17        17         17
Downtown Toronto Stn A           1             1         1          1
East Toronto                     4             4         4          4
East Toronto Business            1             1         1          1
East York                        4             4         4          4
East York/East Toronto           1             1         1          1
Etobicoke                       11            11        11         11
Etobicoke Northwest              1             1         1          1
Mississauga                      1             1         1          1
North York                      24            24        24         24
Queen's Park                     1             1         1          1
Scarborough                     17            17        17         17
West Toronto                     6             6         6          6
York                             5             5         5          5

### Downtown Toronto, North York, and Scarborough have the most neighborhoods.

## Step 3 - Explore and Cluster the Neighborhoods in Toronto

#### Create a map of the neighborhoods in Toronto

In [16]:
#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
!pip install geopy
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Matplotlib and associated plotting modules
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
!pip install sklearn
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
!pip install folium
import folium # map rendering library


#### First set up an instance of geocoder for Toronto

In [17]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Now create the map of Toronto

In [18]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Define the Foursquare API credentials

In [19]:
CLIENT_ID = 'FFLI1QQCW1LJXXOSTNDXAYO32RNNNTMFZL4OOKJGLIHNWJUR' # your Foursquare ID
CLIENT_SECRET = 'LQLDT1TZOE4AK2RRK5EJ2JALLUAIZI5BQYQWFDJS3B25M2XU' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value


#### Define a function to get venues near a neighborhood

In [20]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Gather the venues in each neighborhood into the _toronto_venues_ dataframe

In [21]:
toronto_venues = getNearbyVenues(names = toronto_data['Neighborhood'],
                                 latitudes = toronto_data['Latitude'],
                                 longitudes = toronto_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

#### Check the size of the _toronto_venues_ dataframe, the number of venues in each neighborhood, and the number of distinct venue categories

In [22]:
print(toronto_venues.shape)
toronto_venues.head()

(2152, 7)


Neighborhood Neighborhood Latitude Neighborhood Longitude  \
0         Parkwoods            43.7532586            -79.3296565   
1         Parkwoods            43.7532586            -79.3296565   
2         Parkwoods            43.7532586            -79.3296565   
3  Victoria Village            43.7258823            -79.3155716   
4  Victoria Village            43.7258823            -79.3155716   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1                     KFC       43.754387       -79.333021   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4             Tim Hortons       43.725517       -79.313103   

         Venue Category  
0                  Park  
1  Fast Food Restaurant  
2     Food & Drink Shop  
3          Hockey Arena  
4           Coffee Shop

In [23]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            3   
Alderwood, Long Branch                                               9   
Bathurst Manor, Wilson Heights, Downsview North                     24   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   25   
...                                                                ...   
Willowdale West                                                      5   
Willowdale, Newtonbrook                                              1   
Woburn                                                               3   
Woodbine Heights                                                     7   
York Mills West                                                      3   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             3   
Alderwood, Long Branch                                                9   
Bathurst Manor, Wilson Heights, Downsview North                      24   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    25   
...                                                                 ...   
Willowdale West                                                       5   
Willowdale, Newtonbrook                                               1   
Woburn                                                                3   
Woodbine Heights                                                      7   
York Mills West                                                       3   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            3               3   
Alderwood, Long Branch                               9               9   
Bathurst Manor, Wilson Heights, Downsview North     24              24   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   25              25   
...                                                ...             ...   
Willowdale West                                      5               5   
Willowdale, Newtonbrook                              1               1   
Woburn                                               3               3   
Woodbine Heights                                     7               7   
York Mills West                                      3               3   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      3   
Alderwood, Long Branch                                         9   
Bathurst Manor, Wilson Heights, Downsview North               24   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             25   
...                                                          ...   
Willowdale West                                                5   
Willowdale, Newtonbrook                                        1   
Woburn                                                         3   
Woodbine Heights                                               7   
York Mills West                                                3   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     3  
Alderwood, Long Branch                                        9 

In [24]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 278 uniques categories.


#### Since there are about the same number of unique categories for all of Toronta as there were for all of Manhattan, I'll include all of Toronto's boroughs in my cluster analysis.

#### First, reshape the venue data.

In [25]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
cols = list(toronto_onehot.columns.values)
cols.pop(cols.index('Neighborhood'))
toronto_onehot = toronto_onehot[['Neighborhood'] + cols]

toronto_onehot.head()


Neighborhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2         Parkwoods                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Vegetarian / Vegan Restaurant  \
0                 0                    0  ...                              0   
1                 0                    0  ...                              0   
2                 0                    0  ...                              0   
3                 0                    0  ...                              0   
4                 0                    0  ...                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0         0          0            0              0            0  
1         0          0            0              0            0  
2         0          0            0              0            0  
3         0          0            0              0            0  
4         0          0            0              0            0  

[5 rows x 278 columns]

In [26]:
toronto_onehot.shape

(2152, 278)

#### Next, group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


In [27]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Accessories Store  \
0                                         Agincourt                0.0   
1                            Alderwood, Long Branch                0.0   
2   Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                   Bayview Village                0.0   
4                 Bedford Park, Lawrence Manor East                0.0   
..                                              ...                ...   
95                                  Willowdale West                0.0   
96                          Willowdale, Newtonbrook                0.0   
97                                           Woburn                0.0   
98                                 Woodbine Heights                0.0   
99                                  York Mills West                0.0   

    Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0              0.0      0.0                 0.0           0.0             0.0   
1              0.0      0.0                 0.0           0.0             0.0   
2              0.0      0.0                 0.0           0.0             0.0   
3              0.0      0.0                 0.0           0.0             0.0   
4              0.0      0.0                 0.0           0.0             0.0   
..             ...      ...                 ...           ...             ...   
95             0.0      0.0                 0.0           0.0             0.0   
96             0.0      0.0                 0.0           0.0             0.0   
97             0.0      0.0                 0.0           0.0             0.0   
98             0.0      0.0                 0.0           0.0             0.0   
99             0.0      0.0                 0.0           0.0             0.0   

    Airport Service  Airport Terminal  American Restaurant  ...  \
0               0.0               0.0                 0.00  ...   
1               0.0               0.0                 0.00  ...   
2               0.0               0.0                 0.00  ...   
3               0.0               0.0                 0.00  ...   
4               0.0               0.0                 0.04  ...   
..              ...               ...                  ...  ...   
95              0.0               0.0                 0.00  ...   
96              0.0               0.0                 0.00  ...   
97              0.0               0.0                 0.00  ...   
98              0.0               0.0                 0.00  ...   
99              0.0               0.0                 0.00  ...   

    Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                             0.0               0.0     0.000000   
1                             0.0               0.0     0.000000   
2                             0.0               0.0     0.000000   
3                             0.0               0.0     0.000000   
4                             0.0               0.0     0.000000   
..                            ...               ...          ...   
95                            0.0               0.0     0.000000   
96                            0.0               0.0     0.000000   
97                            0.0               0.0     0.000000   
98                            0.0               0.0     0.142857   
99                            0.0               0.0     0.000000   

    Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                     0.0              0.0       0.0        0.0          0.0   
1                     0.0              0.0       0.0        0.0          0.0   
2                     0.0              0.0       0.0        0.0          0.0   
3                     0.0              0.0       0.0        0.0          0.0   
4                     0.0              0.0       0.0        0.0          0.0   
..                    ...              ...       ...        ...          ..

In [28]:
toronto_grouped.shape

(100, 278)

#### Print each neighborhood with its five most common venues

In [29]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0  Latin American Restaurant  0.33
1                     Lounge  0.33
2             Breakfast Spot  0.33
3         Miscellaneous Shop  0.00
4                      Motel  0.00


----Alderwood, Long Branch----
                venue  freq
0         Pizza Place  0.22
1                 Gym  0.11
2      Sandwich Place  0.11
3  Athletics & Sports  0.11
4         Coffee Shop  0.11


----Bathurst Manor, Wilson Heights, Downsview North----
                venue  freq
0                Bank  0.08
1         Coffee Shop  0.08
2       Grocery Store  0.04
3  Chinese Restaurant  0.04
4         Gas Station  0.04


----Bayview Village----
                 venue  freq
0  Japanese Restaurant  0.25
1                 Café  0.25
2                 Bank  0.25
3   Chinese Restaurant  0.25
4          Music Venue  0.00


----Bedford Park, Lawrence Manor East----
                     venue  freq
0           Sandwich Place  0.08
1              Coffee Shop  0.08
2 

#### Put this information into a new _pandas_ dataframe.  First write a function that puts each neighborhood's data into descending order.  Then use it to create the new dataframe.

In [30]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [57]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                       Bank   
3                                  Bayview Village        Japanese Restaurant   
4                Bedford Park, Lawrence Manor East             Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue    4th Most Common Venue  \
0                Lounge        Breakfast Spot       Miscellaneous Shop   
1                   Gym        Sandwich Place       Athletics & Sports   
2           Coffee Shop         Grocery Store       Chinese Restaurant   
3                  Café                  Bank       Chinese Restaurant   
4           Coffee Shop    Italian Restaurant  Comfort Food Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                 Motel   Moroccan Restaurant   Monument / Landmark   
1           Coffee Shop                   Pub            Playground   
2           Gas Station           Supermarket        Sandwich Place   
3           Music Venue         Movie Theater                 Motel   
4      Toy / Game Store        Breakfast Spot           Pizza Place   

             8th Most Common Venue       9th Most Common Venue  \
0  Molecular Gastronomy Restaurant  Modern European Restaurant   
1                         Pharmacy       Performing Arts Venue   
2               Frozen Yogurt Shop                  Restaurant   
3              Moroccan Restaurant         Monument / Landmark   
4                          Butcher                        Café   

            10th Most Common Venue  
0                Mobile Phone Shop  
1                    Movie Theater  
2              Fried Chicken Joint  
3  Molecular Gastronomy Restaurant  
4                     Liquor Store

#### Now I can do the cluster modeling

#### Run a k-means cluster to sort the data into 5 neighborhoods.

In [58]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1], dtype=int32)

#### Use these clusters to create a new dataframe with the 10 most common venues

In [60]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.shape

(103, 16)

In [61]:
toronto_merged.head(10)

PostalCode           Borough                      Neighborhood    Latitude  \
0        M3A        North York                         Parkwoods  43.7532586   
1        M4A        North York                  Victoria Village  43.7258823   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.6542599   
3        M6A        North York  Lawrence Manor, Lawrence Heights   43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.6623015   
5        M9A         Etobicoke                  Islington Avenue  43.6678556   
6        M1B       Scarborough                    Malvern, Rouge  43.8066863   
7        M3B        North York                   Don Mills North  43.7459058   
8        M4B         East York   Parkview Hill, Woodbine Gardens  43.7063972   
9        M5B  Downtown Toronto          Garden District, Ryerson  43.6571618   

     Longitude  Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
0  -79.3296565             1.0  Fast Food Restaurant       Food & Drink Shop   
1  -79.3155716             0.0           Pizza Place            Hockey Arena   
2  -79.3606359             0.0           Coffee Shop                  Bakery   
3  -79.4647633             0.0        Clothing Store  Furniture / Home Store   
4  -79.3894938             0.0           Coffee Shop        Sushi Restaurant   
5  -79.5322424             NaN                   NaN                     NaN   
6  -79.1943534             0.0  Fast Food Restaurant       Accessories Store   
7   -79.352188             0.0  Caribbean Restaurant                     Gym   
8   -79.309937             0.0           Pizza Place            Intersection   
9  -79.3789371             0.0           Coffee Shop          Clothing Store   

   3rd Most Common Venue      4th Most Common Venue 5th Most Common Venue  \
0                   Park          Accessories Store    Miscellaneous Shop   
1  Portuguese Restaurant                Coffee Shop     French Restaurant   
2                   Park                        Pub        Breakfast Spot   
3      Accessories Store      Vietnamese Restaurant           Event Space   
4            Yoga Studio                       Café              Beer Bar   
5                    NaN                        NaN                   NaN   
6     Miscellaneous Shop                      Motel   Moroccan Restaurant   
7                   Café         Athletics & Sports   Japanese Restaurant   
8               Pharmacy                  Pet Store                  Café   
9         Cosmetics Shop  Middle Eastern Restaurant    Italian Restaurant   

       6th Most Common Venue            7th Most Common Venue  \
0        Moroccan Restaurant              Monument / Landmark   
1          Mobile Phone Shop                            Motel   
2                    Theater                             Café   
3                Coffee Shop                         Boutique   
4        Fried Chicken Joint                              Spa   
5                        NaN                              NaN   
6        Monument / Landmark  Molecular Gastronomy Restaurant   
7  Middle Eastern Restaurant               Miscellaneous Shop   
8                        Spa                             Bank   
9                       Café              Japanese Restaurant   

             8th Most Common Venue       9th Most Common Venue  \
0  Molecular Gastronomy Restaurant  Modern European Restaurant   
1              Moroccan Restaurant         Monument / Landmark   
2                      Yoga Studio              Cosmetics Shop   
3            Performing Arts Venue                        Park   
4               Mexican Restaurant               Smoothie Shop   
5                              NaN                         NaN   
6       Modern European Restaurant           Mobile Phone Shop   
7               Mexican Restaurant           Mobile Phone Shop   
8                        Gastropub          Athletics & Sports   
9               

In [70]:
toronto_merged[['Cluster Labels']].describe()

Cluster Labels
count      100.000000
mean         0.300000
std          0.758787
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          4.000000

#### The _toronto_merged_ dataframe has NaN values in the 'Cluster Labels' column.  Need to replace the NaN values with numpy arrays [0].  For some reason, couldn't get fillna() to work, so had to iterate through rows and check each one.

In [75]:
fill_array = np.array([0.0])

for index, row in toronto_merged.iterrows():
  if np.isnan(row['Cluster Labels']):
    print(str(row['Cluster Labels']))
 
    toronto_merged.loc[index, 'Cluster Labels'] = fill_array


toronto_merged[['Cluster Labels']].describe()

Cluster Labels
count      103.000000
mean         0.291262
std          0.749262
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          4.000000

#### Put the information in the new dataframe onto a map.

In [96]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Examine the clusters, to determine which venues determine the differences between them.

#### Cluster 1 - Coffee shops and pizza places appear to be most common venues in this cluster.

In [97]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0.0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1               North York             0.0           Pizza Place   
2         Downtown Toronto             0.0           Coffee Shop   
3               North York             0.0        Clothing Store   
4             Queen's Park             0.0           Coffee Shop   
5                Etobicoke             0.0                   NaN   
..                     ...             ...                   ...   
96        Downtown Toronto             0.0           Pizza Place   
97        Downtown Toronto             0.0           Coffee Shop   
99        Downtown Toronto             0.0           Coffee Shop   
100  East Toronto Business             0.0    Light Rail Station   
102              Etobicoke             0.0        Hardware Store   

      2nd Most Common Venue  3rd Most Common Venue  4th Most Common Venue  \
1              Hockey Arena  Portuguese Restaurant            Coffee Shop   
2                    Bakery                   Park                    Pub   
3    Furniture / Home Store      Accessories Store  Vietnamese Restaurant   
4          Sushi Restaurant            Yoga Studio                   Café   
5                       NaN                    NaN                    NaN   
..                      ...                    ...                    ...   
96              Coffee Shop                   Café                 Bakery   
97                     Café                  Hotel                    Gym   
99      Japanese Restaurant       Sushi Restaurant                Gay Bar   
100                     Spa          Auto Workshop             Comic Shop   
102  Thrift / Vintage Store           Burger Joint            Social Club   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1       French Restaurant     Mobile Phone Shop                 Motel   
2          Breakfast Spot               Theater                  Café   
3             Event Space           Coffee Shop              Boutique   
4                Beer Bar   Fried Chicken Joint                   Spa   
5                     NaN                   NaN                   NaN   
..                    ...                   ...                   ...   
96                    Pub            Restaurant    Italian Restaurant   
97             Restaurant   Japanese Restaurant           Salad Place   
99             Restaurant           Yoga Studio           Men's Store   
100                  Park      Recording Studio            Restaurant   
102           Flower Shop     Convenience Store  Fast Food Restaurant   

     8th Most Common Venue 9th Most Common Venue  \
1      Moroccan Restaurant   Monument / Landmark   
2              Yoga Studio        Cosmetics Shop   
3    Performing Arts Venue                  Park   
4       Mexican Restaurant         Smoothie Shop   
5                      NaN                   NaN   
..                     ...                   ...   
96               Gastropub            Beer Store   
97      Seafood Restaurant            Steakhouse   
99         Bubble Tea Shop                 Hotel   
100             Skate Park         Burrito Place   
102         Sandwich Place       Supplement Shop   

              10th Most Common Venue  
1    Molecular Gastronomy Restaurant  
2                                Spa  
3           Mediterranean Restaurant  
4                      Burrito Place  
5                                NaN  
..                               ...  
96                      Liquor Store  
97                            Bakery  
99                        Smoke Shop  
100                   Farmers Market  
102                       Kids Store  

[85 rows x 12 columns]

#### Cluster 2 - Parks are either 1st, 2nd, or 3rd most common venues in this cluster.

In [98]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1.0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0               North York             1.0  Fast Food Restaurant   
21                    York             1.0                  Park   
35  East York/East Toronto             1.0     Convenience Store   
49              North York             1.0                  Park   
61         Central Toronto             1.0    Photography Studio   
66              North York             1.0     Electronics Store   
68         Central Toronto             1.0      Sushi Restaurant   
77               Etobicoke             1.0     Mobile Phone Shop   
83         Central Toronto             1.0          Tennis Court   
85             Scarborough             1.0            Playground   
91        Downtown Toronto             1.0                  Park   

   2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
0      Food & Drink Shop                        Park     Accessories Store   
21         Women's Store                         Bar     Accessories Store   
35                  Park           Accessories Store    Miscellaneous Shop   
49                Bakery  Construction & Landscaping      Basketball Court   
61              Bus Line                        Park           Swim School   
66     Convenience Store                        Park     Accessories Store   
68         Jewelry Store                       Trail                  Park   
77                  Park              Sandwich Place     Accessories Store   
83                Lawyer                        Park     Accessories Store   
85          Intersection                        Park     Mobile Phone Shop   
91            Playground                       Trail    Miscellaneous Shop   

   5th Most Common Venue 6th Most Common Venue  \
0     Miscellaneous Shop   Moroccan Restaurant   
21    Miscellaneous Shop   Moroccan Restaurant   
35                 Motel   Moroccan Restaurant   
49     Mobile Phone Shop                 Motel   
61     Accessories Store    Miscellaneous Shop   
66                 Motel   Moroccan Restaurant   
68     Accessories Store   Moroccan Restaurant   
77                 Motel   Moroccan Restaurant   
83    Miscellaneous Shop   Moroccan Restaurant   
85                 Motel   Moroccan Restaurant   
91   Moroccan Restaurant   Monument / Landmark   

              7th Most Common Venue            8th Most Common Venue  \
0               Monument / Landmark  Molecular Gastronomy Restaurant   
21              Monument / Landmark  Molecular Gastronomy Restaurant   
35              Monument / Landmark  Molecular Gastronomy Restaurant   
49              Moroccan Restaurant              Monument / Landmark   
61              Monument / Landmark  Molecular Gastronomy Restaurant   
66              Monument / Landmark  Molecular Gastronomy Restaurant   
68              Monument / Landmark  Molecular Gastronomy Restaurant   
77              Monument / Landmark  Molecular Gastronomy Restaurant   
83              Monument / Landmark  Molecular Gastronomy Restaurant   
85              Monument / Landmark  Molecular Gastronomy Restaurant   
91  Molecular Gastronomy Restaurant       Modern European Restaurant   

              9th Most Common Venue      10th Most Common Venue  
0        Modern European Restaurant           Mobile Phone Shop  
21       Modern European Restaurant           Mobile Phone Shop  
35       Modern European Restaurant           Mobile Phone Shop  
49  Molecular Gastronomy Restaurant  Modern European Restaurant  
61       Modern European Restaurant           Mobile Phone Shop  
66       Modern European Restaurant           Mobile Phone Shop  
68       Modern European Restaurant           Mobile Phone Shop  
77       Modern European Restaurant   Middle Eastern Restaurant  
83       Modern European Restaurant           Mobile Phone Shop  
85       Modern European Restaurant          Miscellaneous Shop  
91                Mobile Phone Shop   Middle Eastern Restaurant

#### Cluster 3 - Pools tend to be most common venues in this cluster.

In [99]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2.0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
57   North York             2.0        Baseball Field   
98    Etobicoke             2.0                  Pool   
101   Etobicoke             2.0                  Pool   

          2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
57            Accessories Store     Mobile Phone Shop                 Motel   
98                        River     Accessories Store     Mobile Phone Shop   
101  Construction & Landscaping        Baseball Field     Accessories Store   

    5th Most Common Venue 6th Most Common Venue  \
57    Moroccan Restaurant   Monument / Landmark   
98                  Motel   Moroccan Restaurant   
101     Mobile Phone Shop                 Motel   

               7th Most Common Venue            8th Most Common Venue  \
57   Molecular Gastronomy Restaurant       Modern European Restaurant   
98               Monument / Landmark  Molecular Gastronomy Restaurant   
101              Moroccan Restaurant              Monument / Landmark   

               9th Most Common Venue      10th Most Common Venue  
57                Miscellaneous Shop                      Museum  
98        Modern European Restaurant          Miscellaneous Shop  
101  Molecular Gastronomy Restaurant  Modern European Restaurant

#### Cluster 4 - Park is 1st most common venue and miscellaneous shops are 2nd or 3rd most common venues in this cluster.

In [100]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3.0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
40  North York             3.0                  Park               Airport   
52  North York             3.0                  Park    Miscellaneous Shop   
64        York             3.0                  Park    Miscellaneous Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
40    Miscellaneous Shop                 Motel   Moroccan Restaurant   
52                 Motel   Moroccan Restaurant   Monument / Landmark   
64                 Motel   Moroccan Restaurant   Monument / Landmark   

              6th Most Common Venue            7th Most Common Venue  \
40              Monument / Landmark  Molecular Gastronomy Restaurant   
52  Molecular Gastronomy Restaurant       Modern European Restaurant   
64  Molecular Gastronomy Restaurant       Modern European Restaurant   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
40  Modern European Restaurant     Mobile Phone Shop      Accessories Store  
52           Mobile Phone Shop     Accessories Store                 Museum  
64           Mobile Phone Shop     Accessories Store                 Museum

#### Cluster 5 - Scarborough is its own cluster

In [101]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4.0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
32  Scarborough             4.0            Playground     Mobile Phone Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
32         Movie Theater                 Motel   Moroccan Restaurant   

   6th Most Common Venue            7th Most Common Venue  \
32   Monument / Landmark  Molecular Gastronomy Restaurant   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
32  Modern European Restaurant    Miscellaneous Shop    Martial Arts School